File: crates/quarry/src/serial.rs
//! Code27 serial math and the 20-char `snextra` helper.
//!
//! A cloud/FIT device is identified by a serial whose **3-char model code** sits
//! at positions 5–7 (0-indexed `4..7`). A controller maps that code to a model,
//! so a converted device must present a serial with the *target* model's code.

use crate::Error;

/// The Code27 alphabet — index `0..27` maps to a check character.
pub const CODE27: &[u8; 27] = b"1DK3R5FPWME47GTX8VL2J6C9NQH";

/// Length of a standard 12-char serial (11 body + 1 check char).
pub const SERIAL_LEN: usize = 12;
/// Length of the u-boot "extra serial" (`snextra`, config field 19).
pub const SNEXTRA_LEN: usize = 20;

/// Compute the Code27 check character over an 11-char body.
///
/// `check = CODE27[(sum of the body's byte values) % 27]`.
pub fn check_char(body: &str) -> Result<char, Error> {
    if body.len() != SERIAL_LEN - 1 {
        return Err(Error::BadLength {
            field: "serial body",
            want: SERIAL_LEN - 1,
            got: body.len(),
        });
    }
    if !body.bytes().all(|b| b.is_ascii_alphanumeric()) {
        return Err(Error::BadChar {
            field: "serial body",
        });
    }
    let sum: usize = body.bytes().map(|b| b as usize).sum();
    Ok(CODE27[sum % 27] as char)
}

/// Build a valid 12-char serial from a 4-char prefix, 3-char model code, and
/// 4-char suffix, appending the Code27 check character.
pub fn make_serial(prefix4: &str, model_code3: &str, suffix4: &str) -> Result<String, Error> {
    if prefix4.len() != 4 {
        return Err(Error::BadLength {
            field: "prefix",
            want: 4,
            got: prefix4.len(),
        });
    }
    if model_code3.len() != 3 {
        return Err(Error::BadLength {
            field: "model_code",
            want: 3,
            got: model_code3.len(),
        });
    }
    if suffix4.len() != 4 {
        return Err(Error::BadLength {
            field: "suffix",
            want: 4,
            got: suffix4.len(),
        });
    }
    let body = format!("{prefix4}{model_code3}{suffix4}");
    let c = check_char(&body)?;
    Ok(format!("{body}{c}"))
}

/// True if `s` is a well-formed 12-char serial with a matching check character.
pub fn validate_serial(s: &str) -> bool {
    if s.len() != SERIAL_LEN {
        return false;
    }
    let (body, tail) = s.split_at(SERIAL_LEN - 1);
    match check_char(body) {
        Ok(c) => tail.starts_with(c),
        Err(_) => false,
    }
}

/// The 3-char model code (positions 5–7) of a serial or `snextra`.
pub fn model_code(s: &str) -> Result<&str, Error> {
    if s.len() < 7 {
        return Err(Error::BadLength {
            field: "serial",
            want: 7,
            got: s.len(),
        });
    }
    Ok(&s[4..7])
}

/// Build a 20-char `snextra` (u-boot config field 19) with the given 3-char model
/// code at positions 5–7, zero-padded. This is the value the cloud firmware reads
/// via `setconfig -g 19` and sends as `sn=` at check-in.
///
/// The prefix defaults to `"SWLW"` when empty. NOTE: `snextra` is a raw 20-char
/// field; the Code27 check character applies to the 12-char serial form, not here.
pub fn make_snextra(prefix: &str, model_code3: &str) -> Result<String, Error> {
    if model_code3.len() != 3 {
        return Err(Error::BadLength {
            field: "model_code",
            want: 3,
            got: model_code3.len(),
        });
    }
    let prefix = if prefix.is_empty() { "SWLW" } else { prefix };
    if prefix.len() != 4 || !prefix.bytes().all(|b| b.is_ascii_alphanumeric()) {
        return Err(Error::BadChar { field: "prefix" });
    }
    let mut s = String::with_capacity(SNEXTRA_LEN);
    s.push_str(prefix);
    s.push_str(model_code3);
    while s.len() < SNEXTRA_LEN {
        s.push('0');
    }
    Ok(s)
}

/// True if `s` is a plausible 20-char `snextra` (correct length, ASCII-alnum,
/// and the model code at 5–7 is extractable).
pub fn validate_snextra(s: &str) -> bool {
    s.len() == SNEXTRA_LEN && s.bytes().all(|b| b.is_ascii_alphanumeric()) && model_code(s).is_ok()
}

#[cfg(test)]
mod tests {
    use super::*;

    // Verified on real hardware this project came from: EPC1X420001 -> check '1'.
    #[test]
    fn check_char_known_vector() {
        assert_eq!(check_char("EPC1X420001").unwrap(), '1');
    }

    #[test]
    fn validate_known_serial() {
        assert!(validate_serial("EPC1X4200011"));
        assert!(!validate_serial("EPC1X4200012")); // wrong check char
        assert!(!validate_serial("EPC1X420001")); // too short
    }

    #[test]
    fn make_and_validate_roundtrip() {
        let s = make_serial("EPC1", "X42", "0001").unwrap();
        assert_eq!(s, "EPC1X4200011");
        assert!(validate_serial(&s));
        assert_eq!(model_code(&s).unwrap(), "X42");
    }

    #[test]
    fn model_code_positions() {
        assert_eq!(model_code("EPC1X4200011").unwrap(), "X42"); // ECW230v3
        assert_eq!(model_code("AAAAX45BBBBc").unwrap(), "X45"); // EWS377-FIT
    }

    #[test]
    fn snextra_shape() {
        let x = make_snextra("EPC1", "X42").unwrap();
        assert_eq!(x.len(), SNEXTRA_LEN);
        assert_eq!(&x[4..7], "X42");
        assert!(validate_snextra(&x));
        assert!(!validate_snextra("EPC1X42")); // too short
        assert!(!validate_snextra("EPC1X420000000000!0")); // bad char
    }

    #[test]
    fn make_serial_length_guards() {
        assert!(make_serial("EPC", "X42", "0001").is_err());
        assert!(make_serial("EPC1", "X4", "0001").is_err());
    }
}

File: go/internal/band/serial.go
// Package band provisions device identity — like ringing a bird.
//
// It mirrors the Rust `quarry` Code27 math in pure Go (so `pelegrun` needs no
// subprocess) and adds a collision-checked local inventory so every physical AP
// gets a unique, valid serial (Constitution: fleet identity is 1:1, fail closed).
package band

import (
	"errors"
	"fmt"
	"strings"
)

// Code27 alphabet: index 0..27 -> check character.
const Code27 = "1DK3R5FPWME47GTX8VL2J6C9NQH"

const (
	SerialLen  = 12 // 11 body + 1 check char
	SnextraLen = 20 // u-boot "extra serial" (config field 19)
)

var (
	ErrLength = errors.New("wrong length")
	ErrChar   = errors.New("disallowed character")
)

func alnum(s string) bool {
	for _, r := range s {
		if !((r >= '0' && r <= '9') || (r >= 'A' && r <= 'Z') || (r >= 'a' && r <= 'z')) {
			return false
		}
	}
	return true
}

// CheckChar computes the Code27 check character over an 11-char body.
func CheckChar(body string) (byte, error) {
	if len(body) != SerialLen-1 {
		return 0, fmt.Errorf("%w: body must be %d chars", ErrLength, SerialLen-1)
	}
	if !alnum(body) {
		return 0, ErrChar
	}
	sum := 0
	for i := 0; i < len(body); i++ {
		sum += int(body[i])
	}
	return Code27[sum%27], nil
}

// MakeSerial builds a valid 12-char serial from prefix(4)+model(3)+suffix(4).
func MakeSerial(prefix4, model3, suffix4 string) (string, error) {
	if len(prefix4) != 4 || len(model3) != 3 || len(suffix4) != 4 {
		return "", fmt.Errorf("%w: need 4+3+4", ErrLength)
	}
	body := prefix4 + model3 + suffix4
	c, err := CheckChar(body)
	if err != nil {
		return "", err
	}
	return body + string(c), nil
}

// ValidateSerial reports whether s is a 12-char serial with a matching check char.
func ValidateSerial(s string) bool {
	if len(s) != SerialLen {
		return false
	}
	c, err := CheckChar(s[:SerialLen-1])
	return err == nil && s[SerialLen-1] == c
}

// ModelCode returns the 3-char model code (positions 5-7) of a serial/snextra.
func ModelCode(s string) (string, error) {
	if len(s) < 7 {
		return "", ErrLength
	}
	return s[4:7], nil
}

// MakeSnextra builds a 20-char field-19 value with model3 at positions 5-7,
// zero-padded. prefix defaults to "SWLW".
func MakeSnextra(prefix, model3 string) (string, error) {
	if len(model3) != 3 {
		return "", fmt.Errorf("%w: model code must be 3 chars", ErrLength)
	}
	if prefix == "" {
		prefix = "SWLW"
	}
	if len(prefix) != 4 || !alnum(prefix) {
		return "", ErrChar
	}
	s := prefix + model3
	return s + strings.Repeat("0", SnextraLen-len(s)), nil
}
